```
======================================================================
ERROR: test_checkpoint_with_multiple_topics (corehq.apps.change_feed.tests.test_kafka_change_feed.KafkaCheckpointTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/change_feed/tests/test_kafka_change_feed.py", line 52, in test_checkpoint_with_multiple_topics
    publish_stub_change(topics.FORM)
  File "/mnt/corehq/apps/change_feed/tests/test_kafka_change_feed.py", line 77, in publish_stub_change
    send_to_kafka(_get_producer(), topic, meta)
  File "/mnt/corehq/apps/change_feed/producer.py", line 26, in send_to_kafka
    _send_to_kafka()
  File "/mnt/corehq/apps/change_feed/producer.py", line 17, in _send_to_kafka
    bytes(json.dumps(change_meta.to_json())),
  File "/vendor/lib/python2.7/site-packages/kafka/producer/keyed.py", line 43, in send_messages
    return self._send_messages(topic, partition, *msg, key=key)
  File "/vendor/lib/python2.7/site-packages/kafka/producer/base.py", line 390, in _send_messages
    fail_on_error=self.sync_fail_on_error
  File "/vendor/lib/python2.7/site-packages/kafka/client.py", line 476, in send_produce_request
    resps = self._send_broker_aware_request(payloads, encoder, decoder)
  File "/vendor/lib/python2.7/site-packages/kafka/client.py", line 166, in _send_broker_aware_request
    payload.partition)
  File "/vendor/lib/python2.7/site-packages/kafka/client.py", line 86, in _get_leader_for_partition
    self.load_metadata_for_topics(topic)
  File "/vendor/lib/python2.7/site-packages/kafka/client.py", line 364, in load_metadata_for_topics
    resp = self.send_metadata_request(topics)
  File "/vendor/lib/python2.7/site-packages/kafka/client.py", line 430, in send_metadata_request
    return self._send_broker_unaware_request(payloads, encoder, decoder)
  File "/vendor/lib/python2.7/site-packages/kafka/client.py", line 130, in _send_broker_unaware_request
    raise KafkaUnavailableError('All servers failed to process request')
KafkaUnavailableError: All servers failed to process request
```

--- a/corehq/ex-submodules/pillowtop/checkpoints/manager.py
+++ b/corehq/ex-submodules/pillowtop/checkpoints/manager.py
@@ -37,10 +37,6 @@ class PillowCheckpoint(object):
         self.checkpoint_id = checkpoint_id
         self._last_checkpoint = None
 
-    def get_or_create(self, verify_unchanged=False):
-        result = self.get_or_create_wrapped(verify_unchanged)
-        return DocGetOrCreateResult(DjangoPillowCheckpoint.to_dict(result.document), result.created)
-
     def get_or_create_wrapped(self, verify_unchanged=False):
         result = get_or_create_checkpoint(self.checkpoint_id)
         checkpoint, created = result

--- a/corehq/ex-submodules/pillowtop/checkpoints/manager.py
+++ b/corehq/ex-submodules/pillowtop/checkpoints/manager.py
@@ -38,10 +38,10 @@ class PillowCheckpoint(object):
         self._last_checkpoint = None
 
     def get_or_create(self, verify_unchanged=False):
-        result = self._get_or_create(verify_unchanged)
+        result = self.get_or_create_wrapped(verify_unchanged)
         return DocGetOrCreateResult(DjangoPillowCheckpoint.to_dict(result.document), result.created)
 
-    def _get_or_create(self, verify_unchanged=False):
+    def get_or_create_wrapped(self, verify_unchanged=False):
         result = get_or_create_checkpoint(self.checkpoint_id)
         checkpoint, created = result
         if (verify_unchanged and self._last_checkpoint and
@@ -60,7 +60,7 @@ class PillowCheckpoint(object):
         pillow_logging.info(
             "(%s) setting checkpoint: %s" % (self.checkpoint_id, seq)
         )
-        checkpoint = self._get_or_create(verify_unchanged=True).document
+        checkpoint = self.get_or_create_wrapped(verify_unchanged=True).document
         checkpoint.sequence = seq
         checkpoint.timestamp = datetime.utcnow()
         checkpoint.save()
@@ -74,7 +74,7 @@ class PillowCheckpoint(object):
         Update the checkpoint timestamp without altering the sequence.
         :param min_interval: minimum interval between timestamp updates
         """
-        checkpoint = self._get_or_create(verify_unchanged=True).document
+        checkpoint = self.get_or_create_wrapped(verify_unchanged=True).document
         now = datetime.utcnow()
         previous = self._last_checkpoint.timestamp if self._last_checkpoint else None
         do_update = True

--- a/corehq/ex-submodules/pillowtop/migrations/0002_djangopillowcheckpoint_sequence_format.py
+++ b/corehq/ex-submodules/pillowtop/migrations/0002_djangopillowcheckpoint_sequence_format.py
@@ -1,0 +1,20 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('pillowtop', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='djangopillowcheckpoint',
+            name='sequence_format',
+            field=models.CharField(default=b'text', max_length=20, choices=[(b'text', b'text'), (b'json', b'json')]),
+            preserve_default=True,
+        ),
+    ]

--- a/corehq/ex-submodules/pillowtop/models.py
+++ b/corehq/ex-submodules/pillowtop/models.py
@@ -1,4 +1,11 @@
+import json
 from django.db import models
+
+
+SEQUENCE_FORMATS = (
+    ('text', 'text'),
+    ('json', 'json'),
+)
 
 
 class DjangoPillowCheckpoint(models.Model):
@@ -7,6 +14,14 @@ class DjangoPillowCheckpoint(models.Model):
     sequence = models.TextField()
     timestamp = models.DateTimeField(auto_now=True)
     old_sequence = models.TextField(null=True, blank=True)
+    sequence_format = models.CharField(max_length=20, choices=SEQUENCE_FORMATS, default='text')
+
+    @property
+    def wrapped_sequence(self):
+        if self.sequence_format == 'json':
+            return json.loads(self.sequence)
+        else:
+            return self.sequence
 
     class Meta:
         app_label = "pillowtop"

